File: CSharp/Sorting/MergeSort.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace CSharp.Sorting
{
    public static class MergeSort
    {
        public static int[] Sort(int[] arrayToSort)
        {
            // BASE CASE: arrays with fewer than 2 elements are sorted
            if (arrayToSort.Length < 2)
            {
                return arrayToSort;
            }

            // STEP 1: divide the array in half
            // We use integer division, so we'll never get a "half index"
            int midIndex = arrayToSort.Length / 2;

            int[] left = MakeSubarray(arrayToSort, 0, midIndex);
            int[] right = MakeSubarray(arrayToSort, midIndex, arrayToSort.Length);

            // STEP 2: sort each half
            int[] sortedLeft = Sort(left);
            int[] sortedRight = Sort(right);

            // STEP 3: merge the sorted halves
            int[] sortedArray = new int[arrayToSort.Length];

            int currentLeftIndex = 0;
            int currentRightIndex = 0;

            for (int currentSortedIndex = 0; currentSortedIndex < arrayToSort.Length;
                    currentSortedIndex++)
            {
                // sortedLeft's first element comes next
                // if it's less than sortedRight's first
                // element or if sortedRight is exhausted
                if (currentLeftIndex < sortedLeft.Length
                        && (currentRightIndex >= sortedRight.Length  // means right array is empty
                        || sortedLeft[currentLeftIndex] < sortedRight[currentRightIndex]))
                {
                    sortedArray[currentSortedIndex] = sortedLeft[currentLeftIndex];
                    currentLeftIndex++;
                }
                else
                {
                    sortedArray[currentSortedIndex] = sortedRight[currentRightIndex];
                    currentRightIndex++;
                }
            }

            return sortedArray;
        }

        public static T[] MakeSubarray<T>(T[] source, int begin, int end)
        {
            if (source == null)
            {
                throw new ArgumentNullException(nameof(source), "Source array is null");
            }

            if (begin > source.Length)
            {
                throw new ArgumentOutOfRangeException(nameof(begin),
                    "Begin index is too big");
            }

            if (end > source.Length)
            {
                throw new ArgumentOutOfRangeException(nameof(end),
                    "End index is too big");
            }

            if (begin > end)
            {
                throw new ArgumentOutOfRangeException(nameof(begin),
                    "Begin index is larger than end index");
            }

            T[] destination = new T[end - begin];
            if (destination.Length > 0)
            {
                Array.Copy(source, begin, destination, 0, destination.Length);
            }

            return destination;
        }
    }
}

File: python/python-algo-ds/chap12/merge_sort.py
def merge_sort(S):
    """Sort (in-place) elements of Python list S using merge-sort"""
    n = len(S)
    if n < 2:
        return

    mid = n // 2
    S1 = S[0:mid]
    S2 = S[mid:n]

    merge_sort(S1)
    merge_sort(S2)

    merge(S1, S2, S)


def merge(S1, S2, S):
    i = j = 0
    while i + j < len(S):
        if j == len(S2) or (i < len(S1) and S1[i] < S2[j]):
            S[i + j] = S1[i]
            i += 1
        else:
            S[i + j] = S2[j]
            j += 1


if __name__ == "__main__":
    l = [4, 5, 6, 34, 2, 7, 8, 9, 0]
    merge_sort(l)
    print(l)
